response_it.go

```diff
@@ -0,0 +1,148 @@
+package tarantool
+
+import (
+	"sync"
+	"time"
+)
+
+// ResponseIterator is an interface for iteration over a set of responses.
+type ResponseIterator interface {
+	// Next tries to switch to a next Response and returns true if it exists.
+	Next() bool
+	// Value returns a current Response if it exists, nil otherwise.
+	Value() *Response
+	// Err returns error if it happens.
+	Err() error
+}
+
+// TimeoutResponseIterator is an interface that extends ResponseIterator
+// and adds the ability to change a timeout for the Next() call.
+type TimeoutResponseIterator interface {
+	ResponseIterator
+	// WithTimeout allows to set up a timeout for the Next() call.
+	// Note: in the current implementation, there is a timeout for each
+	// request: Connection.opts.Timeout. You need to increase the value
+	// if necessary.
+	WithTimeout(timeout time.Duration) TimeoutResponseIterator
+}
+
+type asyncResponseIterator struct {
+	fut     *Future
+	resp    *Response
+	timeout time.Duration
+	cur_pos int
+	err     error
+	done    bool
+}
+
+func (it *asyncResponseIterator) Next() bool {
+	if it.done || it.err != nil {
+		return false
+	}
+
+	var (
+		resp  *Response
+		err   error
+		final bool
+	)
+
+	if it.timeout == 0 {
+		it.fut.mutex.RLock()
+		defer it.fut.mutex.RUnlock()
+
+		resp = it.nextResponse()
+		if it.resp == nil {
```

Comment:
I don't get:
- if iterator doesn't get timeout parameter
- then it doesn't wait anything
- therefore usually it doesn't get anything, because nextResponse sees empty `fut.pushes` and empty `resp`.
LOL
What i'm missing?

Reply:
Exactly, see Lua implementation:
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/push/
Lua's iterator just leave a loop if nothing to do (returns it.Next() == false in our case).
```
-- There is no pause because conn:call does not wait for
-- server_function to finish. The first time that we go through
-- the pairs() loop, we see the messages table is empty. Like this:
-- tarantool> messages
```

Reply:
Go is not Lua. It doesn't need to match Lua's code exactly.
In Go one gets notifications through channels. Channels are composable.
"Polling" on `it.Next()` would be definitely anti-pattern in Go.

Well... I could not came with clear api now.
It doesn't look like iterator could be efficiently implemented in Go-friendly way if there is no separate method "conn.CallStream".